File: plugin/src/main/java/org/apache/grails/intellij/plugin/projectView/nodes/OtherGrailsAppSourcesNode.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package org.apache.grails.intellij.plugin.projectView.nodes;

import com.intellij.ide.projectView.PresentationData;
import com.intellij.ide.projectView.ViewSettings;
import com.intellij.ide.projectView.impl.nodes.PsiDirectoryNode;
import com.intellij.ide.util.treeView.AbstractTreeNode;
import com.intellij.openapi.project.Project;
import com.intellij.openapi.vfs.VirtualFile;
import com.intellij.psi.PsiDirectory;
import com.intellij.psi.PsiFile;
import com.intellij.psi.PsiManager;
import com.intellij.util.PlatformIcons;
import org.jetbrains.annotations.NotNull;
import org.apache.grails.intellij.plugin.artefact.api.ArtefactHandlers;
import org.apache.grails.intellij.plugin.artefact.api.GrailsDisplayableArtefactHandler;
import org.apache.grails.intellij.plugin.projectView.NodeWeights;
import org.apache.grails.intellij.plugin.projectView.impl.GrailsViewItems;
import org.apache.grails.intellij.plugin.structure.GrailsApplication;

import java.util.ArrayList;
import java.util.Collection;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Objects;
import java.util.Set;

public class OtherGrailsAppSourcesNode extends GrailsPsiDirectoryNode {

  private volatile GrailsApplication grailsApplication;

  public OtherGrailsAppSourcesNode(@NotNull PsiDirectory directory, @NotNull ViewSettings settings) {
    super(directory, settings, PlatformIcons.SOURCE_FOLDERS_ICON, NodeWeights.OTHER_GRAILS_APP_FOLDER, null);
  }

  public @NotNull GrailsApplication getGrailsApplication() {
    GrailsApplication result = grailsApplication;
    if (result == null) {
      result = TreeNodeUtil.findNotNullValueOfType(this, GrailsApplication.class);
      grailsApplication = result;
    }
    return result;
  }

  @Override
  public @NotNull Collection<AbstractTreeNode<?>> getChildrenImpl() {
    GrailsApplication application = getGrailsApplication();
    Project project = application.getProject();
    PsiManager manager = PsiManager.getInstance(project);
    Set<String> specialNames = GrailsViewItems.SPECIAL_GRAILS_APP_FOLDERS.keySet();

    // directories from nodes that are showed separately
    Set<VirtualFile> visibleArtefactDirs = new LinkedHashSet<>();
    for (GrailsDisplayableArtefactHandler handler : ArtefactHandlers.displayableArtefactHandlers()) {
      if (!handler.isVisible(application)) continue;
      VirtualFile directory = handler.getDirectory(application);
      if (directory != null) visibleArtefactDirs.add(directory);
    }
    List<VirtualFile> artefactDirs = new ArrayList<>();
    for (VirtualFile dir : visibleArtefactDirs) {
      if (!specialNames.contains(dir.getName())) artefactDirs.add(dir);
    }

    // other grails-app directories excluding special, special are showed separately
    List<VirtualFile> otherDirs = new ArrayList<>();
    for (VirtualFile child : application.getAppRoot().getChildren()) {
      if (!artefactDirs.contains(child) && !specialNames.contains(child.getName())) otherDirs.add(child);
    }

    Collection<AbstractTreeNode<?>> result = new ArrayList<>();
    // do not show artefact nodes under these directories
    for (VirtualFile dir : artefactDirs) {
      PsiDirectory directory = manager.findDirectory(dir);
      if (directory != null) {
        result.add(new PsiDirectoryNode(project, directory, getSettings(), GrailsViewItems::shouldShowItem));
      }
    }
    for (VirtualFile dir : otherDirs) {
      PsiDirectory directory = manager.findDirectory(dir);
      if (directory != null) {
        result.add(new PsiDirectoryNode(project, directory, getSettings()));
      }
    }
    return result;
  }

  @Override
  public boolean contains(@NotNull VirtualFile file) {
    if (!super.contains(file)) return false;
    PsiFile psiFile = PsiManager.getInstance(Objects.requireNonNull(getProject())).findFile(file);
    return psiFile != null && GrailsViewItems.shouldShowItem(psiFile);
  }

  @Override
  protected void updateImpl(@NotNull PresentationData data) {
    super.updateImpl(data);
    data.setLocationString("Other sources");
  }
}

File: src/org/jetbrains/plugins/grails/projectView/nodes/OtherGrailsAppSourcesNode.kt
/*
 * Copyright 2000-2026 JetBrains s.r.o. and contributors.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.jetbrains.plugins.grails.projectView.nodes

import com.intellij.ide.projectView.PresentationData
import com.intellij.ide.projectView.ViewSettings
import com.intellij.ide.projectView.impl.nodes.PsiDirectoryNode
import com.intellij.ide.util.treeView.AbstractTreeNode
import com.intellij.openapi.vfs.VirtualFile
import com.intellij.psi.PsiDirectory
import com.intellij.psi.PsiManager
import com.intellij.util.PlatformIcons
import com.intellij.util.lazyUnsafe
import org.jetbrains.plugins.grails.artefact.api.displayableArtefactHandlers
import org.jetbrains.plugins.grails.projectView.NodeWeights
import org.jetbrains.plugins.grails.projectView.impl.shouldShowItem
import org.jetbrains.plugins.grails.projectView.impl.specialGrailsAppFolders
import org.jetbrains.plugins.grails.structure.GrailsApplication

class OtherGrailsAppSourcesNode(
  directory: PsiDirectory,
  settings: ViewSettings
) : GrailsPsiDirectoryNode(
  directory,
  settings,
  nodeWeight = NodeWeights.OTHER_GRAILS_APP_FOLDER,
  nodeIcon = PlatformIcons.SOURCE_FOLDERS_ICON
) {

  val grailsApplication: GrailsApplication by lazyUnsafe { findNotNullValueOfType<GrailsApplication>() }

  override fun getChildrenImpl(): Collection<AbstractTreeNode<*>> {
    val application = grailsApplication
    val project = application.project
    val manager = PsiManager.getInstance(project)

    // directories from nodes that are showed separately
    val artefactDirs = displayableArtefactHandlers.filter {
      it.isVisible(application)
    }.mapNotNullTo(HashSet()) {
      it.getDirectory(application)
    }.filter {
      it.name !in specialGrailsAppFolders.keys
    }

    // other grails-app directories excluding special, special are showed separately
    val otherDirs = application.appRoot.children.filter {
      it !in artefactDirs && it.name !in specialGrailsAppFolders.keys
    }

    // do not show artefact nodes under these directories
    val artefactDirNodes = artefactDirs.mapNotNull {
      manager.findDirectory(it)
    }.map { PsiDirectoryNode(project, it, settings, ::shouldShowItem) }

    val otherDirNodes = otherDirs.mapNotNull {
      manager.findDirectory(it)
    }.map { PsiDirectoryNode(project, it, settings) }

    return artefactDirNodes + otherDirNodes
  }

  override fun contains(file: VirtualFile): Boolean = super.contains(file) && PsiManager.getInstance(project!!).findFile(file)?.let {
    shouldShowItem(it)
  } ?: false

  override fun updateImpl(data: PresentationData) {
    super.updateImpl(data)
    data.apply {
      locationString = "Other sources"
    }
  }
}